### PR TITLE
Create a parser to parse the model from a json

### PR DESCRIPTION
**Pull request by sreenikSS**
The json file should contain the loss function, init type, optimizer details and the network layers corresponding to the architecture in mind. The format of the json is mentioned in the issue https://github.com/mlpack/mlpack/issues/1254

There are a number of flaws, I will try to remove them in subsequent commits. I shall address them in the issue itself.

### CODE REVIEW

**Comment by zoq** [comment]
Thanks for opening the PR, do you mind to change the title of the PR to something more descriptive?

**Comment by sreenikSS** [comment]
Oops, I'm sorry for that. Fixed it

File: src/mlpack/methods/ann/modelParser.cpp [hunk 1]
```diff
@@ -0,0 +1,1132 @@
...
+
+/**
+ * @author Eugene Freyman
+ * Returns the accuracy (percentage of correct answers).
+ * @param predLabels predicted labels of data points.
+ * @param realY real labels (they are double because we usually read them from
+ * CSV file that contain many other double values).
+ * @return percentage of correct answers.
+ */
+double accuracy(arma::Row<size_t> predLabels, const arma::mat& realY)
```
Comment (MuLx10) [inline src/mlpack/methods/ann/modelParser.cpp]: It is better to include the file instead of writing it again

File: src/mlpack/methods/ann/modelParser.cpp [hunk 2]
```diff
@@ -0,0 +1,1132 @@
...
+using namespace mlpack::optimization;
+using namespace arma;
+using namespace std;
+using namespace boost::property_tree;
+
+/**
+ * Implementation of the entire dataset consisting of the training data
+ * and validation data
+ */
+class Dataset
```
Comment (MuLx10) [inline src/mlpack/methods/ann/modelParser.cpp]: Would you mind splitting the file into .hpp and .cpp

File: src/mlpack/methods/ann/modelParser.cpp [hunk 3]
```diff
@@ -0,0 +1,1132 @@
...
+/**
+ * Update the values of a given stl map with that of another map 
+ * corresponding to the keys that are common
+ * 
+ * Keys are of type string and values are of type double
+ * 
+ * @param origParams The map whose values will be updated
+ * @param newParams The map whose values will be used to update origParams
+ */
+void updateParams(map<string, double> &origParams, map<string, double> &newParams)
```
Comment (MuLx10) [inline src/mlpack/methods/ann/modelParser.cpp]: You can use a copy constructor instead or directly assign values to optimizerparam instead of copying in loop.

File: src/mlpack/methods/ann/modelParser.cpp (hunk 1 above)
Comment (sreenikSS) [reply src/mlpack/methods/ann/modelParser.cpp]: Thanks for the suggestion, but the file is in the *models* repo, so I cannot really include it. Moreover it is basic and simple code, so I am not sure if code repetition can be a big issue here.

File: src/mlpack/methods/ann/modelParser.cpp (hunk 2 above)
Comment (sreenikSS) [reply src/mlpack/methods/ann/modelParser.cpp]: Sure I will do it. Actually the code has some visible scope of improvement that I am aware of (of course, along with things I am not aware of). I will certainly split it before finally requesting a merge. It might have some redundancy in the include statements as well. I'll clean them up after my exams end (or if I can manage time in between).

File: src/mlpack/methods/ann/modelParser.cpp (hunk 3 above)
Comment (sreenikSS) [reply src/mlpack/methods/ann/modelParser.cpp]: Thanks for the suggestion. That might be a better solution, though I will have to look into it a bit more. One reason why it does what it does is that I wanted to check for errors. For example, some keys are absolutely necessary, some are not. That validation part is not done yet. I will keep this in mind while implementing the validation process.

File: src/mlpack/methods/ann/modelParser.cpp [hunk 4]
```diff
@@ -0,0 +1,1132 @@
...
+
+/** The final implementation of this file would not have a main method. This is 
+ * merely to ease testing. The following include satement can hence be 
+ * removed later
+*/
+#include <mlpack/core/data/split_data.hpp>
+int main()
+{
+  //arma::mat trainX, trainY;
+  string fileName = "network3.json";
```
Comment (zoq) [inline src/mlpack/methods/ann/modelParser.cpp]: Perhaps you can post this file somewhere? Is the json file autogenerated?

Comment (sreenikSS) [reply src/mlpack/methods/ann/modelParser.cpp]: Sorry for the delay. I had completely missed this comment. 
The json file is user specified. I think it will be a lot clearer if you take a look [here](https://github.com/sreenikSS/Testing/blob/master/Dataset3.json)

**Comment by rcurtin** [comment]
Hey @sreenikSS, still interested in working on this one?  We can let it go stale (or let someone else pick it up) if not. :+1:

**Comment by sreenikSS** [comment]
@rcurtin I'd guess letting it go stale would be fine because as per our previous discussions, this is supposed to be part the new *model converter* repository when it'll be created.

As far as anyone else taking this issue up is concerned, I don't think that will be a good idea because the work is complete (and lying around in my own repo) :)
